src/processor/tree/tree_wrapper.rs

```diff
@@ -79,35 +96,61 @@ impl TreeWrapper {
             hex::encode(root_hash)
         );
 
-        assert_eq!(root_hash.as_bytes(), block.new_state_root);
-        tracing::debug!("Successfully processed block {}", block.l2_block_number);
+        let root_hash_bytes = root_hash.as_bytes();
+        if root_hash_bytes == block.new_state_root {
+            tracing::debug!("Successfully processed block {}", block.l2_block_number);
+
+            Ok(root_hash)
+        } else {
+            tracing::error!(
+                "Root hash mismatch: {:?} vs. {:?}",
+                root_hash_bytes,
+                block.new_state_root
+            );
+            let mut rollback_entries = Vec::with_capacity(self.index2key.len());
+            for (index, key) in &self.index2key {
+                let value = self.key2value.get(key).unwrap();
+                rollback_entries.push(TreeEntry::new(*key, *index, *value));
+            }
 
-        root_hash
+            self.tree.extend(rollback_entries);
```

Comment:
How does this interact with `MerkleTree::latest_version`? If it increments the version count whilst keeping the tree the same, we'll probably run into some issues with the compacted value processing

Reply:
hmm, which issues specifically?

Reply:
but yes, this does leave invalid state in the Merkle tree history, and if the result passes our validity checks, it probably means they're somehow lacking...